---
title: "cleaned_comm_dyn_metaG_code"
output: rmarkdown::github_document
date: "2024-10-28"
---

##### 1. Load Packages
```{r}
library(ggplot2)
library(ggsci)
library(dplyr)
library(scales)
library(pheatmap)
library(viridis)
library(tidyplots)
library(readxl)
```

##### 2. Define Functions for QC Checks
```{r}

##### 1. Plot Total DNA Reads in each sequenced sample #####
plot_total_dna_reads<- function(input_dna_reads_file) {
  
  # Load File #
  fastqc <- read.csv(file = input_dna_reads_file)
  fastqc$sequences_millions <- as.numeric(fastqc$sequences_millions)
  
  # Plot Data #
  fastqc_check<- ggplot(fastqc, aes(x=Sample.Name,y=sequences_millions)) + 
    geom_bar(stat="identity",position = 'dodge',color='white',fill='#4682B4')+
    theme_classic()+
    theme(axis.text.x = element_blank(),
          axis.text.y = element_text(family = "Helvetica",color='black',size=28),
          axis.title.x = element_blank(),
          axis.title.y = element_text(family = "Helvetica",color='black',size=28),
          legend.text = element_text(family = "Helvetica",size = 28),
          legend.title = element_text(family = "Helvetica",size = 28),
          strip.text = element_text(color='Black', size = 28),
          panel.border=element_blank(),
          legend.position='top')+
    ylab("Number of reads (million)")+
    geom_hline(yintercept = 1, color = "black", linetype = "solid", linewidth = 1) +
    scale_y_continuous(expand = c(0, 0))
  return(fastqc_check)
}


##### 2. Plot number of internal DNA standards recovered #####
plot_dna_standards<- function(dna_standards_file) {
  # Load File #
  dna_standards <- read.csv(file = dna_standards_file)
  dna_standards_graph<- ggplot(dna_standards, aes(x=sample_name,y=`genome_equivalents`)) + 
    geom_bar(stat="identity",position = 'dodge',color='black',fill='#CBC3E3')+
    theme_classic()+
    theme(axis.text.x = element_text(angle = 90,hjust=0.9,vjust=0.5,family = "Helvetica",color='black',size=16),
          axis.text.y = element_text(family = "Helvetica",color='black',size=22),
          axis.title.x = element_text(family = "Helvetica",color='black',size=22),
          axis.title.y = element_text(family = "Helvetica",color='black',size=22),
          legend.text = element_text(family = "Helvetica",size = 22),
          legend.title = element_text(family = "Helvetica",size = 22),
          strip.text = element_text(color='Black', size = 22),
          panel.border=element_blank(),
          legend.position='top')+
    xlab("")+
    ylab("genome equivalents")+
    scale_y_continuous(expand = c(0, 0))
  return(dna_standards_graph)
}

##### 2. Plot number of internal DNA standards recovered #####
plot_metaG_FCM_comparison<- function(fcm_counts_file) {
  # Load File #
  fcm_counts <- read.csv(file = fcm_counts_file)
  
  fcm_scatter<-ggplot(fcm_counts, aes(x=new_name,y=het_ratio,fill=experiment,color=experiment))+
    geom_line(aes(group=experiment),size=1)+
    geom_point(size=3)+
    #geom_errorbar(aes(ymin = mean_absolute_counts - se_y, ymax = mean_absolute_counts + se_y), width = 0.06,size=0.5)+  # Error bar layer
    theme_classic()+
    theme(axis.text.x = element_text(angle = 90,hjust=0.9,vjust=0.5,family = "Helvetica",color='black',size=16),
          #axis.text.x = element_text(family = "Helvetica",color='black',size=16),
          axis.text.y = element_text(family = "Helvetica",color='black',size=22),
          axis.title.x = element_text(family = "Helvetica",color='black',size=22),
          axis.title.y = element_text(family = "Helvetica",color='black',size=22),
          legend.text = element_text(family = "Helvetica",size = 18),
          legend.title = element_text(family = "Helvetica",size = 18),
          strip.text = element_text(color='Black', size = 22),
          panel.border=element_blank(),
          legend.position='top')+
    xlab("")+
    ylab("relative abundance of heterotrophs")
    #facet_grid(cols=vars(experiment_day),scales="free",space="free")
    #scale_color_manual(values = alpha(c("#ff3145","#eb91e3","#4f2c7f",
    #                                    "#007bc0","#00e674","#ff7f00","#cbbbc2")))+
    #scale_y_log10(limits = c(50000, 100000000))+
    #annotation_logticks(sides="l")
  return(fcm_scatter)
}

```
##### 2. Generate QC Plots
```{r,fig.width=10, fig.height=6}
##### Define Input Files #####
input_dna_reads_file='/Users/jamesmullet/dna_community_dynamics_data.csv'
dna_standards_file='/Users/jamesmullet/dna_standard_reads_mapped.csv'
fcm_counts_file = '/Users/jamesmullet/fcm_metaG_comparison.csv'

##### 1. Generate Total DNA Reads Plot #####
dna_counts_plot<- plot_total_dna_reads(input_dna_reads_file)
dna_counts_plot

##### Generate Number of Recovered DNA Standards #####
internal_standard_plot <- plot_dna_standards(dna_standards_file)
internal_standard_plot

##### FCM vs metaG Comparison #####
fcm_metaG_plot <- plot_metaG_FCM_comparison(fcm_counts_file)
fcm_metaG_plot
```

##### 3. Define Functions for Read Mapping Results
```{r}

##### 1. Plot the Genome Equivalents in the community over time - Stacked Bar Chart #####
plot_read_map_results<- function(read_map_results_file,cluster_colors) {
  input_read_mapping <- read.csv(file = read_map_results_file)
  input_read_mapping <- input_read_mapping%>% filter(experiment_name != 'Thermus_standard')
  metaG_plot<- ggplot(input_read_mapping, aes(x=replicate,y=`absolute_counts`,fill=experiment_name)) + 
    geom_bar(position="stack", stat="identity",alpha=0.8,color='black')+
    #geom_bar(position="fill", stat="identity",alpha=0.9,color='black')+
    #geom_bar(stat="identity",position = 'dodge',color='black')+
    theme_classic()+
    theme(axis.text.x = element_text(family = "Helvetica",color='black',size=16),
          axis.text.y = element_text(family = "Helvetica",color='black',size=16),
          axis.title.x = element_text(family = "Helvetica",color='black',size=16),
          axis.title.y = element_text(family = "Helvetica",color='black',size=16),
          legend.text = element_text(family = "Helvetica",size = 16),
          legend.title = element_text(family = "Helvetica",size = 16),
          strip.text = element_text(color='Black', size = 16),
          panel.border=element_blank(),
          legend.position='right')+
    xlab("")+
    #ylab("relative abundance (%)")+
    ylab("absolute genome equivalent counts")+
    scale_fill_manual(values = cluster_colors,guide = guide_legend(ncol = 1, byrow = TRUE))+
    #scale_fill_manual(values = alpha(c("#ff3145","#eb91e3","#4f2c7f","#007bc0","#00e674","#ff7f00","#cbbbc2")))+
    #scale_fill_lancet()+
    facet_grid(cols=vars(sampling_day),scales="free",space="free")+
    #geom_hline(yintercept = 1, color = "black", linetype = "solid", size = 1) +
    scale_y_continuous(limits = c(0, 750000000),expand = c(0, 0))
  return(metaG_plot)
}

##### 2. Plot the Genome Equivalents in the community over time - Line Plot #####
plot_read_map_results_lineplot<- function(read_map_results_file,cluster_colors) {
  input_read_mapping <- read.csv(file = read_map_results_file)
  averaged_data <- input_read_mapping %>%
    group_by(sampling_day,experiment_name,groupings) %>%
    summarise(mean_absolute_counts = mean(absolute_counts),
              se_y = sqrt(var(absolute_counts) / n()),
              .groups = 'drop') %>%
    as.data.frame()
  
  averaged_data <- averaged_data%>% filter(experiment_name != 'Thermus_standard')

  scatterplot_version<-ggplot(averaged_data, aes(x=sampling_day,y=mean_absolute_counts,fill=experiment_name,color=experiment_name))+
    geom_line(aes(group=experiment_name),size=1)+
    geom_point(size=3)+
    geom_errorbar(aes(ymin = mean_absolute_counts - se_y, ymax = mean_absolute_counts + se_y), width = 0.06,size=0.5)+  # Error bar layer
    theme_classic()+
    theme(axis.text.x = element_text(family = "Helvetica",color='black',size=16),
          axis.text.y = element_text(family = "Helvetica",color='black',size=16),
          axis.title.x = element_text(family = "Helvetica",color='black',size=16),
          axis.title.y = element_text(family = "Helvetica",color='black',size=16),
          legend.text = element_text(family = "Helvetica",size = 16),
          legend.title = element_text(family = "Helvetica",size = 16),
          strip.text = element_text(color='Black', size = 16),
          panel.border=element_blank(),
          legend.position='top')+
    xlab("")+
    #ylab("relative abundance (%)")+
    ylab("log10 absolute genome equivalent counts")+
    scale_color_manual(values = cluster_colors)+
    guides(color = guide_legend(nrow = 2))+
    scale_y_log10(limits = c(2000000, 400000000))+
    annotation_logticks(sides="l")
  
    #facet_grid(rows=vars(experiment_name),scales = 'free_y')
    #scale_y_continuous(limits = c(0, 30000000),labels = scales::comma_format(scale = 1e-7))
  return(scatterplot_version)
}

##### 3. Extraction Efficiency Comparison #####
lineplot_extraction_efficiency<- function(read_map_results_file,cluster_colors) {
  input_read_mapping <- read.csv(file = read_map_results_file)
  
  extract_input <- input_read_mapping%>% filter(sampling_day == '0')
  extract_input$initial_FCM_cell_count_added <- as.numeric(extract_input$initial_FCM_cell_count_added)
  extract_input$extraction_efficiency <- (extract_input$genome_equivalents / extract_input$initial_FCM_cell_count_added)
  
  
  averaged_extraction_efficiency <- extract_input %>%
    group_by(sampling_day,experiment_name) %>%
    summarise(mean_extraction_efficiency = mean(extraction_efficiency),
              se_y = sqrt(var(extraction_efficiency) / n()),
              .groups = 'drop') %>%
    as.data.frame()
  mini_averaged_extraction_efficiency <- subset(averaged_extraction_efficiency, select = c(experiment_name,mean_extraction_efficiency))
  
  merged_efficiency_df <- merge(input_read_mapping, mini_averaged_extraction_efficiency, by = "experiment_name")
  
  # Calculate relative counts
  merged_efficiency_df <- merged_efficiency_df %>%
    mutate(corrected_genome_recovery = genome_recovery / genome_recovery[sampling_day == '0'])
  
  merged_efficiency_df$corrected_extraction_efficiency <- (merged_efficiency_df$mean_extraction_efficiency * merged_efficiency_df$corrected_genome_recovery)
  
  merged_efficiency_df$extraction_efficiency_corrected_counts <- (merged_efficiency_df$genome_equivalents / merged_efficiency_df$corrected_extraction_efficiency)
  
    
  averaged_data <- merged_efficiency_df %>%
    group_by(sampling_day,experiment_name,groupings) %>%
    summarise(mean_absolute_counts = mean(extraction_efficiency_corrected_counts),
              se_y = sqrt(var(extraction_efficiency_corrected_counts) / n()),
              .groups = 'drop') %>%
    as.data.frame()
  
  merged_efficiency_df <- merged_efficiency_df%>% filter(experiment_name != 'Thermus_standard')

  scatterplot_version <- merged_efficiency_df %>%
      tidyplot(x=sampling_day,y=extraction_efficiency_corrected_counts,color=experiment_name,dodge_width = 0)%>%
      add_mean_line(linewidth=0.6) %>%
      add_mean_dot(size=2) %>%
      add_sem_errorbar(dodge_width = NULL,width=0.6,linewidth=0.4) %>%
      adjust_y_axis(title = "Estimated Cell Count",transform = "log10") %>%
      #adjust_y_axis(title = "Estimated Cell Count") %>%
      adjust_x_axis(title = "Time (Days)") %>%
      adjust_colors(new_colors = cluster_colors) %>%
      adjust_size(width = 80, height = 50, unit = "mm") %>%
      adjust_legend_title("Organism")
  return(scatterplot_version)
}



```

##### 3. Generate Read Mapping Plots
```{r}
##### Define Input Files #####
read_map_results_file='/Users/jamesmullet/community_dynamics_data_readmap_normalized_absolute_counts.csv'
input_read_mapping <- read.csv(file = read_map_results_file)

##### Define Color Scheme #####
cluster_colors <- c(`1904_Alteromonas`= '#ff3145',`1907_Thalassospira`= '#eb91e3',`1940_Pseudohoeflea`='#4f2c7f',
                    `1943_Marinobacter`= '#007bc0',`Thermus_standard`= '#ff7f00',`MED4`='#00e674')


##### 1. Generate Total DNA Reads Plot #####
read_map_plot<- plot_read_map_results(read_map_results_file,cluster_colors)
read_map_plot

##### 2. Generate Metagenomic Distribution - without extraction efficiency #####
read_map_lineplot<- plot_read_map_results_lineplot(read_map_results_file,cluster_colors)
read_map_lineplot

##### 3. Generate Metagenomic Distribution - accounting for Extraction Efficiency #####
extraction_efficiency_corrected_lineplot<- lineplot_extraction_efficiency(read_map_results_file,cluster_colors)
extraction_efficiency_corrected_lineplot

##### Save Files as SVGs #####
#Cairo(file = "/Users/jamesmullet/comm_dyn_metaG_overview_barchart.svg", type = "svg", width = 40, height = 20, units = "in", dpi = 250)
#print(read_map_plot)  # 'final_graph' is the ggplot object
#dev.off()

#Cairo(file = "/Users/jamesmullet/comm_dyn_metaG_overview_lineplot.svg", type = "svg", width = 40, height = 20, units = "in", dpi = 250)
#print(read_map_lineplot)  # 'final_graph' is the ggplot object
#dev.off()

#Cairo(file = "/Users/jamesmullet/comm_dyn_metaG_extraction_efficiency.svg", type = "svg", width = 40, height = 25, units = "in", dpi = 200)
#print(extraction_efficiency_plot)  # 'final_graph' is the ggplot object
#dev.off()

```
